Stop requestslib from overriding auth headers with data from .netrc file

susodapop opened the pull request:
<!-- Please review our pull request review process in CONTRIBUTING.md before your proceed. -->

Resolves #337

<!---
  Include the number of the issue addressed by this PR above if applicable.
  
  Example:
    resolves #1234

  Please review our pull request review process in CONTRIBUTING.md before your proceed.
-->

### Description

This pull request refactors the credentials handling in `python_submissions.py` using the workaround suggested in #337. The desired behaviour from this change is not obvious when looking only at the code diff. However the idea here is to explicitly order Python requests to use Bearer authentication with an explicitly provided token. If this is not done explicitly, there is an edge case where a local `.netrc` file can override an auth token present in `profiles.yml`.

There's no automated test for this because I'm not sure how to deposit a .netrc file in a test environment.

I reproduced the issue manually by adding this to `~/.netrc` on my workstation:

```
machine <my-workspace>.cloud.databricks.com
login token
password <expired_token>
```

I removed the skip marker on `test_python_uc_databricks_sql_endpoint` and ran this integration test on the `main` branch. The test failed with this exception:

```
ERROR    stdout_log:eventmgr.py:66 [0m18:58:04     b'{"error_code":"403","message":"Invalid access token."}'
ERROR    file_log:eventmgr.py:66 [0m13:58:04.325701 [error] [MainThread]:    b'{"error_code":"403","message":"Invalid access token."}'
```

Then I applied the change in this pull request and re-ran the test. It passed.

<!--- Describe the Pull Request here -->

### Checklist

- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt-databricks next" section.

andrefurlan-db reviewed (changes requested):
it needs to work with oauth as well. Token is not passed in config

susodapop commented:
I've updated the code to use the `HeaderFactory` provided by `databricks.sdk` and verified it still works with PAT auth. Need to do a quick test with OAuth and maybe basic auth to make sure those cases are also handled.

susodapop commented:
I found that the same issue we're fixing here (.netrc overrides auth headers in requests made by dbt-databricks) also affects our databricks-sdk-py repo (.netrc overrides auth headers in requests made for oauth handshake).

The fix in this PR will fix dbt-databricks .netrc woes if the auth method is PAT. But it will only fix the issue for oauth use-cases after databricks-sdk-py is updated with this patch https://github.com/databricks/databricks-sdk-py/issues/107 and we update the dbt-databricks dependency to incorporate that fix.

Note: I found the same bug in `databricks-sql-connector`'s custom implementation of OAuth handshakes _but this doesn't affect dbt-databricks because the dbt adapter uses the OAuth implementation in the SDK_.

susodapop commented:
Now that databricks-sdk has pushed the parallel fix for oauth I've rebased on `main`. This included a refactor of `SessionCredentials` base on a breaking change in databricks-sdk.

susodapop commented:
Looks like there is now a parallel PR that bumps the databricks-sdk version and includes the `RefreshableCredentials` refactor: https://github.com/databricks/dbt-databricks/pull/349

andrefurlan-db reviewed (dismissed):
Left one comment that can be addressed in a separate PR

andrefurlan-db reviewed (commented):
now that I think about it, is there a chance of the token expiring while polling and executing?

Maybe it is better to not set self.auth and get the headers before each request from self._credential_provider ?

benc-db commented:
@susodapop what is this status of this PR?  Also, if we are going to bring this in, we really do need to make sure we're sending the right user agent, since all of the tracking for this library depends on it.

susodapop commented:
It was deprioritised at the time because were chasing down OOM issues in dbt-databricks. We should get this mergeable and release it, although the bug radius is pretty small since it requires `.netrc` to be present and Python models to be in use.

lennartkats-db commented:
@susodapop @andrefurlan-db any update on this? I just ran into this issue as well. This issue should be fairly common among Databricks customers since we still have lots of docs and community content calling out that `.netrc` should be used. #337 causes materialized views to fail with `Error getting info for materialized view/streaming table: <catalog.schema.table>`

benc-db commented:
Looks like there is a conflict to resolve.

benc-db commented on dbt/adapters/databricks/python_submissions.py:
```diff
@@ -31,11 +47,12 @@ def __init__(self, parsed_model: Dict, credentials: DatabricksCredentials) -> No
-        self.auth_header = {
-            "Authorization": f"Bearer {self.credentials.token}",
+        self.extra_headers = {
             "User-Agent": f"dbt-labs-dbt-spark/{DBT_SPARK_VERSION} (Databricks)",
```
This looks like the wrong user agent @susodapop lol.

susodapop replied:
Fixed!

susodapop commented:
My fault. I failed to pull the latest `main` before synchronising. Will fix locally and push again.

susodapop commented:
Alright I've pushed all the changes from code review. I'm going to manually test this with a bad `.netrc` locally and wait for the integration tests to pass before requesting a merge.

susodapop commented:
I have verified that the python models tests fail without my change and a bad `.netrc` is present. With my fix applied, the python model tests work.

susodapop commented:
Alright, I've pushed the final version of this code. I moved the `BearerAuth` class into `connections.py` because we're about to use it there to fix an issue with streaming tables / materialized views. 

I also simplified the code by setting the `Session.auth` property rather than setting the `.auth` property for every single network request.
